runtime/src/accounts_db.rs

```diff
@@ -921,6 +949,96 @@ impl AccountsDB {
         reclaims
     }
 
+    fn squash_uncleaned_keys(&self, max_slot: Slot) {
+        let (keys, mut slots) = self.collect_uncleaned_keys_to_slot(max_slot);
+        if slots.is_empty() {
+            return;
+        }
+
+        slots.sort_unstable();
+        let max_slot = *slots.last().unwrap();
+        let mut squashed_keys = self.squashed_uncleaned_keys.write().unwrap();
+        for key in keys.into_iter().flatten() {
+            squashed_keys.0.insert(key);
+        }
+        squashed_keys.1 = max_slot;
+    }
+
+    fn collect_uncleaned_keys_to_slot(&self, max_slot: Slot) -> (Vec<Vec<Pubkey>>, Vec<Slot>) {
+        let slots: Vec<Slot> = self
+            .uncleaned_keys
+            .iter()
+            .filter_map(|entry| {
+                let slot = entry.key();
+                if *slot <= max_slot {
+                    Some(*slot)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        (
+            slots
+                .iter()
+                .filter_map(|slot| {
+                    let keys = self.uncleaned_keys.remove(slot).unwrap().1;
```

Comment:
ok, lack of drain_filter and avoidance of `.clone()` causing this involved coding... ;)